include/deal.II/base/mutable_bind.h

```diff
@@ -0,0 +1,320 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_base_mutable_bind_h
+#define dealii_base_mutable_bind_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/patterns.h>
+#include <deal.II/base/std_cxx14/utility.h>
+#include <deal.II/base/std_cxx17/tuple.h>
+
+#include <tuple>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Utilities
+{
+  /**
+   * A mutable version of std::bind, that binds all arguments of a function
+   * pointer to a stored tuple, and allows you to update the tuple between
+   * calls.
+   *
+   * An example usage of this class is through the helper function
+   * mutable_bind() that creates a MutableBind object on the fly, based on its
+   * arguments:
+   *
+   * @code
+   * void my_function(const int &a, const double &b);
+   *
+   * auto bound = mutable_bind(my_function, 1, 2.0);
+   *
+   * bound(); // will execute my_function(1, 2.0);
+   *
+   * bound.set_parameters(2, 3.0);
+   * bound(); // will execute my_function(2, 3.0);
+   *
+   * bound.parse_parameters("3: 4.0");
+   * bound(); // will execute my_function(3, 4.0);
+   * @endcode
+   *
+   * The arguments are copied to the tuple, with their reference and const
+   * attributes removed. Only copy constructible objects are allowed as
+   * function arguments. If you need to keep some references around, you may
+   * wrap your function into a std::function, and capture your reference:
+   *
+   * @code
+   * Triangulation<dim> &tria;
+   * std::function<void(double, double)> cube =
+   *    [&tria] (double left, double right) {
+   *    GridGenerator::hyper_cube(tria, left, right);
+   * }
+   *
+   * auto bound_cube = mutable_bind(cube);
+   *
+   * bound_cube.parse_parameters("0: 1");
+   * bound_cube(); // calls GridGenerator::hyper_cube(tria, 0, 1);
+   * @endcode
+   *
+   * @authors Luca Heltai, Matthias Maier, 2019.
+   */
+  template <typename ReturnType, class... FunctionArgs>
+  struct MutableBind
+  {
+    /**
```

Comment:
Is there a particular reason why this is a `struct` instead of a `class`, and why all members are `public`?

Reply:
This you should ask @tamiko :D

Reply:
If there isn't a good reason, then I'd suggest making it a `class` and make most of its member `private` :-)

Reply:
The reason is practical: this is a very simple object, that holds a function and a tuple. As a struct, it is very easy to assign to it: 
```
void f(double d);
MutableBind<void, double> exp = {f, {}};
```
I'd like to keep this possibility.

Reply:
Hm, but that assumes that you know the layout. And it means that you can never *change* the layout either. If you like this assignment type, it would be easiest to just write a constructor for it.

Reply:
Done. Left three constructors.